### DISCUSSION TITLE
Getting "UnsupportedOperation" error on sending data to AWS AppSync

### DISCUSSION THREAD

**Discussion by neha-2022** [artilleryio/artillery#1574] [Artillery]
Configured web socket connection using "connect" and able to create successful web socket connection.
When tried to send subscription registration messages to AWS AppSync using Amazon Cognito user pools authentication, getting below error - 

_ws WS receive: {"type":"error","payload":{"errors":[{"errorType":"UnsupportedOperation","message":"unknown not supported through the realtime channel"}]}}_

Found the solution (https://github.com/apollographql/apollo-feature-requests/issues/224) on the internet but no sure how to use it with artillery.

Version info: 
```
<Artillery Core: 2.0.0-12>
```

Running this command: 
```
<artillery run .\test-script.yml>
```
I expected to see this happen:

Subscription should be registered.

Getting this error while sending subscription registration message to AWS AppSync -

_ws WS receive: {"type":"error","payload":{"errors":[{"errorType":"UnsupportedOperation","message":"unknown not supported through the realtime channel"}]}}_

Files being used:
```
<test-script.yml>
config:
  target: "wss://endpoint/graphql/realtime"
  ensure: 
    maxErrorRate: 1
  phases:
    - duration: 1
      arrivalRate: 1
      name: "sample load run"
  ws:
    headers: 
       'sec-websocket-protocol': "graphql-ws"
  processor: "./processor.js" 
before:
  flow:
    - function: "getToken"
    - think: 3
scenarios:
  - name: " Load testing subscription "
    engine: "ws"
    flow:    
      - connect: "{{ target }}?header={{ encoded_api_header }}&payload=e30="
      - send:
              {
                "id":"1",
                "payload": {
                      "query": "subscription onCreateMessage {
                          onCreateMessage {
                            __typename
                            }
                          }
                        }",
                      "variables": {},
                      "extensions" : {
                          "authorization" : {
                              "Authorization" : "JWT TOKEN",
                              "host": "API_DOMAIN"
                            }
                        }
                },
                "type": "start",
                capture: {
                    json : '$',
                    as : 'response'
                }
              }

```

**Comment by neha-2022** [comment]
Hi @hassy, could you please guide/advise me on above issue.
Thanks a lot!!

**Comment by hassy** [comment]
@neha-2022 I'm not familiar with AppSync but it looks like it implements a custom application-level protocol on top of WebSockets (which is only used for transport).

You'd need to build a custom AppSync engine for Artillery using the official [AppSync client for JS](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-appsync/index.html). The Kinesis engine would be a good example to start with: https://github.com/artilleryio/artillery-engine-kinesis

**Reply by hassy** [reply]
You could try this engine as well:

https://github.com/zoheirkhonyagar/artillery-engine-graphql-ws#readme

**Reply by neha-2022** [reply]
@hassy @zoheirkhonyagar - Yes, I tried using _artillery-engine-graphql-ws_ but it doesn't seem to work. 
WebSocket connection itself isn't established and the script never ends.

<test-script.yml>
config:
  target: "wss://endpoint/graphql/realtime"
  ensure: 
    maxErrorRate: 1
  phases:
    - duration: 1
      arrivalRate: 1
      name: "sample load run"
  ws:
    headers: 
       'sec-websocket-protocol': "graphql-ws"
  engines:
    'graphql-ws' : {}
  processor: "./processor.js" 
before:
  flow:
    - function: "getToken"
    - think: 3
scenarios:
  - name: " Load testing subscription "
    engine: "ws"
    flow:    
      - send:
			{
              id: '1',
              payload: {
                  'data': '{"query":"subscription onCreateMessage {\n onCreateMessage \n }","variables":{}}',
                  'extensions': {
                      'authorization': {
                          'Authorization': "JWT TOKEN",
                          'host': "API_DOMAIN"
                      }
                  }
              },
              type: 'start',
              capture: {
                  json : '$',
                  as : 'response'
                }
            }
